pie/database/__init__.py

```diff
@@ -24,42 +26,165 @@ def __init__(self):
+
+
 def init_core():
     """Load core models and create their tables.
 
     This function is responsible for creation of all core tables.
     """
-    # Everything depends on config, we have to initiate it first
-    importlib.import_module("pie.database.config")
+    # Everything depends on DB versioning, we have to make sure it's created
     database.base.metadata.create_all(database.db)
+    session.commit()
 
-    for module in ("acl", "i18n", "logger", "storage", "spamchannel"):
-        import_stub: str = f"pie.{module}.database"
-        try:
-            importlib.import_module(import_stub)
-            print(
-                f"Database models {COLOR.green}{import_stub}{COLOR.none} imported."
-            )  # noqa: T001
-        except Exception as exc:
-            print(
-                f"Database models {COLOR.red}{import_stub}{COLOR.none} failed: "
-                f"{COLOR.cursive}{exc}{COLOR.none}."
-            )  # noqa: T001
-            raise
+    _import_database(module_name="pie.database.config")
 
-    database.base.metadata.create_all(database.db)
-    session.commit()
+    for module in ("acl", "i18n", "logger", "storage", "spamchannel"):
+        module_name: str = f"pie.{module}.database"
+        _import_database(module_name=module_name)
 
 
 def init_modules():
     """Load all database models and create their tables.
 
     This function is responsible for creation of all module tables.
     """
-    _import_database_tables()
+    repositories: List[str] = _list_directory_directories("modules")
+    for repository in repositories:
+        modules: List[str] = _list_directory_directories(repository)
+        for module in modules:
+            # Detect module's database files
+            database_stub: str = os.path.join(module, "database")
+            if not os.path.isfile(database_stub + ".py") and not os.path.isdir(
+                database_stub
+            ):
+                continue
 
-    database.base.metadata.create_all(database.db)
-    session.commit()
+            module_name = database_stub.replace(os.sep, ".")
+
+            # Import the module DB
+            _import_database(module_name=module_name)
+
+
+def _get_module_tables(module_name: str):
+    return [
+        class_reg_value.__tablename__
+        for class_reg_value in database.base.registry._class_registry.values()
+        if getattr(class_reg_value, "__module__") == module_name
+        and hasattr(class_reg_value, "__tablename__")
+    ]
+
+
+def _import_database(module_name: str):
+    """This functions imports the DB file (if exists).
+
+    It also checks the DB file version against the DBVersion table.
+
+    If the DB scheme is outdated, performs an update.
+
+    Default version is 1 (to stay compatible with)
+
+    :module_name: Name of the module (with dots as path separators)
+    """
+    try:
+        db_module = importlib.import_module(module_name)
+
+        module_version: int = getattr(db_module, "VERSION", 1)
```

Comment:
applies to comment before

Reply:
This is a different thing - if a module DB file does not contain the VERSION attribute, it will assume that the version is 1, which is a way to ensure backward compatibility.